src/segment_reshape/topology/find_related.py

```diff
@@ -111,281 +112,219 @@ def find_related_features(
+            yield QgsGeometry(part.clone())
 
-    main_feature_geom = main_feature.geometry()
-    segment_start_point = clone_geometry_safely(
-        main_feature_geom.vertexAt(main_feature_segment[0])
-    )
-    segment_end_point = clone_geometry_safely(
-        main_feature_geom.vertexAt(main_feature_segment[1])
+
+def _get_containing_component(
+    geom: QgsGeometry, contains_geom: QgsGeometry
+) -> QgsGeometry:
+    for component in _as_point_or_line_components(geom):
+        if component.contains(contains_geom):
+            return component
+
+    raise ValueError(
+        f"could not find component of {geom} that contains {contains_geom}"
     )
 
-    # Find geometries having their boundary intersecting main feature segment
-    common_segment_candidates: List[QgsGeometry] = []
-    for _, feature in related_features_by_layer:
-        geom = feature.geometry()
-        if (
-            geom.type() == QgsWkbTypes.GeometryType.PolygonGeometry
-            and (geom.touches(segment_start_point) and geom.touches(segment_end_point))
-        ) or (
-            geom.type() != QgsWkbTypes.GeometryType.PolygonGeometry
-            and (
-                geom.intersects(segment_start_point)
-                and geom.intersects(segment_end_point)
-            )
-        ):
-            common_segment_candidates.append(geom)
-
-    segments = []
-    end_points = []
-
-    common_segment_as_geom = QgsGeometry()
-    start_point = QgsPoint()
-    end_point = QgsPoint()
-
-    # Return just boundary as common segment if no candidates found
-    if len(common_segment_candidates) == 0:
-        if main_feature_geom.type() == QgsWkbTypes.GeometryType.PointGeometry:
-            return (None, [], [])
-
-        vertices = []
-        if main_feature_geom.type() == QgsWkbTypes.GeometryType.PolygonGeometry:
-            # TODO: multigeometries and holes in polygons
-            vertices = main_feature_geom.asPolygon()[0]
-
-        if main_feature_geom.type() == QgsWkbTypes.GeometryType.LineGeometry:
-            # TODO: multigeometries
-            vertices = main_feature_geom.asPolyline()
-
-        # TODO: How to handle common segments along the boundary
-        # (outside triggered segment)? Return just the boundary for now
-        return (
-            QgsLineString(vertices),
-            [
-                ReshapeCommonPart(
-                    main_feature_layer,
-                    main_feature,
-                    list(range(len(vertices))),
-                    is_reversed=False,
-                )
-            ],
-            [],
-        )
 
-    else:
-        common_segment = _calculate_common_segment(
-            main_feature_geom,
-            common_segment_candidates,
-            segment_start_point,
-            segment_end_point,
-        )
+def _get_line_component_for_part_and_ring(
+    geom: QgsGeometry, part_index: int, ring_index: int
+) -> QgsGeometry:
+    for i, part in enumerate(geom.constParts()):
```

Comment:
would it be simpler to use part_index instead of for loop
`
part = list(geom.constParts())[part_index] # check that part_index is in list
`

Reply:
fixed, also catches now all IndexErrors either from part index or interior ring index failures